src/main/java/com/marklogic/spark/reader/HostUtil.java

```diff
@@ -0,0 +1,66 @@
+package com.marklogic.spark.reader;
+
+import com.marklogic.client.DatabaseClient;
+import com.marklogic.client.datamovement.Forest;
+import com.marklogic.client.datamovement.ForestConfiguration;
+import com.marklogic.client.datamovement.impl.DataMovementManagerImpl;
+import com.marklogic.client.row.RowManager;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Determines hosts for distributing load. A "data host" (or "d-node") is one with at least forest associated with the
+ * database that the given Java Client points to.
+ * <p>
+ * Mostly copied from BatcherImpl and RowBatcherImpl in the Java Client.
```

Comment:
This is the copy/paste job from the Java Client. It would be nice to make that code reusable, particularly the logic for removing a client when it stops responding. That would be a significant refactor though because the logic is buried a bit within DMSDK.